pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +236,89 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut); err != nil {
+		return err
+	}
+
+	if !o.Watch {
+		return nil
+	}
+	return o.WatchAndSync()
+}
+
+// WatchAndSync sets up a recursive filesystem watch on the sync path
+// and invokes rsync each time the path changes
+func (o RsyncOptions) WatchAndSync() error {
+
+	// these variables must be accessed while holding the changeLock
+	// mutex as they are shared between goroutines to communicate
+	// sync state/events.
+	var (
+		changeLock sync.Mutex
+		dirty      bool
+		lastChange time.Time
+		watchError error
+	)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					if e := watcher.Remove(event.Name); e != nil {
+						glog.Warningf("error removing watch for %s: %v", event.Name, e)
+					}
+				} else {
+					watchError = fsnotification.AddRecursiveWatch(watcher, event.Name)
```

Comment:
I think we only want to change `watchError` if it is `nil`?
Otherwise sequential events from the `watcher.Events` channel could lead to overwriting a non-nil `watchError` with a nil value before the check in the `for` loop below comes in, effectively swallowing an error.

If instead of `watchError error` we had a channel of errors, such intricate cases would not happen :)